### ISSUE TITLE
skip shipping and payement problems

### ISSUE THREAD

**Issue by bruno-ds** [Sylius/Sylius#7943]
Hello,

I use both your new features [skip shipping](https://github.com/Sylius/Sylius/pull/7276) and [skip payement](https://github.com/Sylius/Sylius/pull/7031).

(I don't know if you prefer that I put the comment grouped here or use some other way, so please tell me if I'm doing wrong)

here are the issues I encountered  : 
# 1. as said in the original issue [#7215](https://github.com/Sylius/Sylius/issues/7215#issuecomment-294118756) : 

> - if the shipping is skipped, the address step still ask for the shipping address, it should only ask for the billing address. (it could be simpler for you to invert the two steps)
> - if either the shipping and the payement steps are going to be skipped, no address at all should be asked (this is subject to discussion, but in our shop, we decided to go that way)

# 2. the order state remain "new" instead of becoming fulfilled
when an order skip both shipping and payement, its state do not change : it remain "new"
PS : I tested also with just skipping the shipping, the problem still occurs.

it is **really important** since this break this kind of custom logic and I don't see a workaround : 
```yaml
  sylius_order:
    callbacks:
      after:
        app_store_order_complete:
          on: ["fulfill"]
          do: ["@app.event_listener.order_state_machine", "onOrderStateMachineStateComplete"]
          args: ["object"]
```

# 3. when you skip the paiement, 
the state remain "awaiting_payment", it should be "payment_skipped" (I'm not the only on to think that : https://github.com/Sylius/Sylius/pull/7031#issuecomment-271103631)

# 4. same for the shipping, 
the state should become "shipping_skipped", in addition, this state is missing in `Sylius\Component\Core\OrderCheckoutStates`

**Comment by psihius** [comment]
Also, when you enable "skip shipping step" - it does not recalculate the shipping method based on selected country - it ships to the default country it calculated when adding to the cart.
In my personal case the default shipping country is Latvia, on the order checkout I select Estonia and after completing the checkout, the shipping option is still for Latvia.
When I disable the skip shipping step, when I change the country to "Estonia" and come to the selecting the shipping page, the only option is not selected by default in the form. This indicates to me, that it does not re-select the only available option for shipping somewhere in the processing of the order.
This might be the same for the payment options too, by the way - warrants a check.

**Comment by lchrusciel** [comment]
Hey @bruno-ds,

According to your issues, we cannot invert addressing and shipping steps, as shipping methods depend on the address. But some event listener should do the job to build a dynamic address form. Another problem is that we are using shipping address as a `main address`. Removing the shipping one can complicate a few things. 

For sure we need to change a state of an order according to skipped steps, but I'm not sure if additional steps are required. To be 100% correct we should add them, also to be able to trace them. But maybe we could simplify our model and just assume, that if the payment amount equal to 0  then the order is paid and skipped shipping means that it is will be shipped just after completing the payment? The order also should be in `fulfilled` state if both payment and shipping were skipped. 

According to the problem that @psihius spotted it is because of the `DefaultShippingMethodResolver`. The default implementation takes into the consideration just first enabled method available for consideration. Because the customer skipped shipping step, the default option, Latvia, has been left. It is possible to choose a default shipping method based on an address if specified. Would you like to fix it?

**Comment by bruno-ds** [comment]
Hello @lchrusciel , and thank you for your detailed response.


My manager just answered me, we will move in two steps : 
### first I'll try to hardcode a controller action that bypass your implementation 
by setting the good value, something like that (it's not the right code, it's an example of what sylius permits) : 
```php
    public function orderNowAction()
    {
        $order = $this->createRandomOrder();

        $this->addressOrderRandomly($order);
        $this->proceedCheckout($order, OrderCheckoutTransitions::TRANSITION_ADDRESS);

        $this->pickRandomShippingForOrder($order);
        $this->proceedCheckout($order, OrderCheckoutTransitions::TRANSITION_SELECT_SHIPPING);

        $this->pickRandomPaymentForOrder($order);
        $this->proceedCheckout($order, OrderCheckoutTransitions::TRANSITION_SELECT_PAYMENT);

        $order->setNotes('Thank you for the order');
        $this->proceedCheckout($order, OrderCheckoutTransitions::TRANSITION_COMPLETE);

        return $this->redirectToRoute('sylius_shop_homepage');
    }
```

### secondly if still needed, I'll try to fix your code and push a PR
since our priority is to deliver asap in production, this would happen after, in a few months.
Or, if the first step is a failure, this would happen sooner ;) 

PS : in both cases, I cant/won't fix the bug reported by @psihius because we never ship (we sell software)

**Comment by lchrusciel** [comment]
@bruno-ds I'm afraid that your configuration can end up with bugs, that won't be so easy to spot because some of the business logic is handled via state machine callbacks. 
I would recommend you, to customise your state machine and add some shipping and payment state resolver in `sylius_order_payment` and `sylius_order_shipment`. With this solution, you could have decoupled logic for both transitions. Another customization will be required in `sylius_order` state machine to mark the order as fulfilled. Of course, you can go your way, but then you need to cover the behaviour of checkout with tests really well, to ensure, that everything works as expected. Choose wisely :D 

@psihius Can you open your another issue with your problem, just to keep threads clean?

**Comment by psihius** [comment]
@lchrusciel done #7965

**Comment by bruno-ds** [comment]
the dev. is done, here is the main code : 
```php
public function InstantOrderForFreeProducts(int $productVariantId)
    {
        $customer = $this->getCustomer();
        $productVariant = $this->getProductVariant($productVariantId);

        if (null == $productVariant) {
            return $this->createNotFoundException();
        }

        if ($this->preventDuplicate($productVariant, $customer)) {
            return $this->render('AppBundle:SpecialOrders:preventDuplicate.html.twig', [
                'productVariant'  => $productVariant,
            ]);
        }

        $order = $this->createOrder($customer, $productVariant);

        $this->addressOrder($order, $customer);//we use a SSO system, who know the country of the user, the other is set as "n/a"
        $this->proceedOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_ADDRESS);

        if ($this->isTransitionPossibleForOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_SELECT_SHIPPING)) {
            $this->pickRandomShippingForOrder($order);
            $this->proceedOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_SELECT_SHIPPING);
        }

        if ($this->isTransitionPossibleForOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_SELECT_PAYMENT)) {
            $this->pickRandomPaymentForOrder($order);
            $this->proceedOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_SELECT_PAYMENT);
        }

        if ($this->isTransitionPossibleForOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_COMPLETE)) {
            $order->setNotes('Thank you for the order');
            $this->proceedOrderCheckout($order, OrderCheckoutTransitions::TRANSITION_COMPLETE);
        }


        if ($this->isTransitionPossibleForOrder($order, OrderTransitions::TRANSITION_FULFILL)) {
            $this->proceedOrder($order, OrderTransitions::TRANSITION_FULFILL);

            return $this->render('AppBundle:SpecialOrders:InstantOrderForFreeProducts.html.twig', [
                'order'  => $order,
                'productVariant'  => $productVariant,
            ]);

        }
        
        return $this->render('AppBundle:SpecialOrders:Error.html.twig', [
            'order'           => $order,
            'productVariant'  => $productVariant,
        ]);

    }
```

this code is possible thanks to this beta2 change `Sylius\Component\Core\StateResolver\CheckoutStateResolver::resolve()` called by the state machine it automaticaly change the shipping|payment states as skipped if they are elligibles. My only modification comparitively to a checkout beeing made by a customer is : 
```php
        if ($this->isTransitionPossibleForOrder($order, OrderTransitions::TRANSITION_FULFILL)) {
            $this->proceedOrder($order, OrderTransitions::TRANSITION_FULFILL);
//...
        }
```

in sylius' class `Sylius\Component\Core\StateResolver\OrderStateResolver::canOrderBeFulfilled()` there is  this little difference who prevent the order to be fullfilled : 
```php
    private function canOrderBeFulfilled(OrderInterface $order)
    {
        return
            OrderPaymentStates::STATE_PAID === $order->getPaymentState() &&
            OrderShippingStates::STATE_SHIPPED === $order->getShippingState()
        ;
    }
```

I think that the PR will need to at least change this accordingly with new STATE_SKIPPED states for both order and payment.

**Comment by lchrusciel** [comment]
Your bypass just could be achieved in state machine transitions. But if it works for you, I'm ok with that.

What's about skipped states, is that no matter if you skip payment or shipment you still have this objects on your order. Skipping any of them has a sense for me, but unfortunately, it would require some core changes and it won't be trivial. Right now, because you will always have payment and shipment defined on a cart you should make a transition to their final state (`paid` and `skipped`). More complex but better solution will be to introduce `skipped` states as you propose and adjust order processor to not add payment nor shipment if not needed.

**Comment by bruno-ds** [comment]
hello @lchrusciel , and thank you for your time.

**1/** You're right, I could have used a trigger on the `checkout_state` `completed` and used this trigger to change the order `state` to `fulfilled`. But since I wanted to skipp also the address step, it was a better choice for my unconventional use case to bypass everything. 
While there is no "bug correction", I think that another folk should consider your suggestion who is much better than my code.
(just to be clear, my use case is special because at launch, our store will temporarily only have free software, so no shipping, no payment, no address needed)


**2/** reading your answer I try to understand what is a feedback over what I've done, and what is about the sylius' code who need some improvements.
For example the `skipped` states do not exist for the moment so, as you say, it's not possible to use them.
You suggest to make a transition to the existing final states `paid` and `shipped`, I suppose that you suggest me to do so in my bypass( the function [`InstantOrderForFreeProducts`](https://github.com/Sylius/Sylius/issues/7943#issuecomment-296115074) I pasted above). I'm afraid to do so because it would mean doing more stuff untested. For now I do only one state change more than sylius beta 2 do.
In short words : Doing anything more should be done within a PR, no?
